crates/esthri/src/ops/upload.rs

```diff
@@ -0,0 +1,435 @@
+/*
+ * Copyright (C) 2021 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+use std::{borrow::Cow, collections::HashMap, path::Path};
+
+use bytes::{Bytes, BytesMut};
+use futures::{future, stream, Future, Stream, StreamExt, TryStreamExt};
+use log::{debug, info, warn};
+use log_derive::logfn;
+use parking_lot::Mutex;
+use tokio::{
+    fs::File,
+    io::{AsyncRead, AsyncReadExt},
+};
+
+use crate::{
+    bufferpool::{Buffer, Pool},
+    compression::compress_to_tempfile,
+    config::Config,
+    errors::{Error, Result},
+    handle_dispatch_error,
+    rusoto::*,
+};
+
+static PENDING_UPLOADS: Mutex<Vec<PendingUpload>> = parking_lot::const_mutex(Vec::new());
```

Comment:
If I'm understanding the current impl we only store one pending multipart upload and we could have multiple during a sync. This should cancel all of them